File: backend/cara/api/v1/family.py
"""Family endpoints — currently presence (who's home) via frigate-faces."""

from __future__ import annotations

from dataclasses import asdict
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from cara.api.deps import get_current_user
from cara.models.user import User
from cara.services.family import FamilyPresenceUnavailable, people_present

router = APIRouter(prefix="/family", tags=["family"])


class PersonOut(BaseModel):
    name: str
    last_seen: datetime
    minutes_ago: int


class PresenceOut(BaseModel):
    window_minutes: int
    count: int
    people: list[PersonOut]


@router.get("/who-is-home", response_model=PresenceOut)
async def who_is_home(
    window_minutes: Annotated[int, Query(ge=1, le=1440)] = 15,
    _user: User = Depends(get_current_user),  # noqa: B008
) -> PresenceOut:
    try:
        rows = await people_present(window_minutes=window_minutes)
    except FamilyPresenceUnavailable as exc:
        raise HTTPException(status.HTTP_503_SERVICE_UNAVAILABLE, str(exc)) from exc
    return PresenceOut(
        window_minutes=window_minutes,
        count=len(rows),
        people=[PersonOut(**asdict(p)) for p in rows],
    )

File: frontend/src/api/family.ts
import { authFetch } from './auth';

const API = '/api/v1';

export interface PersonAtHome {
  name: string;
  last_seen: string;     // ISO timestamp
  minutes_ago: number;
}

export interface PresenceResponse {
  window_minutes: number;
  count: number;
  people: PersonAtHome[];
}

export async function whoIsHome(windowMinutes = 15): Promise<PresenceResponse> {
  const r = await authFetch(`${API}/family/who-is-home?window_minutes=${windowMinutes}`);
  if (!r.ok) {
    if (r.status === 503) throw new Error('Riconoscimento facciale non disponibile');
    throw new Error(`HTTP ${r.status}`);
  }
  return r.json();
}
